File: lib/node/signing.ts
import crypto from "crypto";

export function b64url(buf: Buffer) {
  return buf.toString("base64").replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/g, "");
}

export function signRequest(opts: {
  method: "POST";
  path: string;        // e.g. /v1/acquiring/qr/create
  clientId: string;
  requestId: string;
  requestTimeMs: string;
  bodyJson: string;    // compact JSON
  privateKeyPem: string;
  keyVersion?: string;
}) {
  const { method, path, clientId, requestId, requestTimeMs, bodyJson, privateKeyPem } = opts;
  const content = `${method}.${path}.${clientId}.${requestId}.${requestTimeMs}.${bodyJson}`;
  const signer = crypto.createSign("RSA-SHA256");
  signer.update(Buffer.from(content, "utf8"));
  const sig = signer.sign(privateKeyPem);
  const signature = encodeURIComponent(b64url(sig));
  const keyVersion = opts.keyVersion ?? "1";
  return `algorithm=RSA256,keyVersion=${keyVersion},signature=${signature}`;
}

File: lib/dotnet/Signing.cs
using System;
using System.Security.Cryptography;
using System.Text;
using System.Web;

public static class Signing {
  static string B64Url(byte[] b) => Convert.ToBase64String(b).Replace("+","-").Replace("/","_").TrimEnd('=');
  
  public static string Sign(string method, string path, string clientId, string requestId, string requestTimeMs, string bodyJson, string privateKeyPem, string keyVersion="1") {
    var content = $"{method}.{path}.{clientId}.{requestId}.{requestTimeMs}.{bodyJson}";
    
    using var rsa = RSA.Create();
    // Starting from .NET Core 3.0, ImportFromPem is available.
    // For older versions, you would need a library like BouncyCastle or parse the PEM manually.
    rsa.ImportFromPem(privateKeyPem);
    
    var sig = rsa.SignData(Encoding.UTF8.GetBytes(content), HashAlgorithmName.SHA256, RSASignaturePadding.Pkcs1);
    
    // Use Uri.EscapeDataString for URL encoding, which is safer than HttpUtility.UrlEncode
    var s = Uri.EscapeDataString(B64Url(sig));
    
    return $"algorithm=RSA256,keyVersion={keyVersion},signature={s}";
  }
}
